### PR TITLE
feat: automate adding new go-libp2p versions to perf

### PR DESCRIPTION
**Pull request by galargh**
This PR implements the process described here: https://github.com/libp2p/go-libp2p/pull/2442#discussion_r1280741230

Additionally:
- it finds the latest go-libp2p version itself
- updates Go in the go.mod file to the latest stable version

Automatically created PR: https://github.com/libp2p/test-plans/pull/255
Automatically triggered perf run: https://github.com/libp2p/test-plans/actions/runs/5836370405

### CODE REVIEW

File: .github/workflows/add-new-impl-versions.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+name: Add new implementation versions
+
+on:
+  push:
```
Comment (galargh) [inline .github/workflows/add-new-impl-versions.yml:4]: I used push for testing. We should change that. With workflow dispatch, it'll have to be triggered manually. It should work just fine if we put it on a schedule too.

```suggestion
  workflow_dispatch:
```

File: perf/impl/go-libp2p/v0.28/.gitignore [hunk 2]
```diff
@@ -1,3 +1,2 @@
 perf
 .cache
-v0.28
```
Comment (galargh) [inline perf/impl/go-libp2p/v0.28/.gitignore:3]: I couldn't find anywhere where this would get created.

File: perf/impl/go-libp2p/v0.28/Makefile [hunk 3]
```diff
@@ -3,11 +3,10 @@ GO_FILES := $(wildcard *.go)
 all: perf
 
 perf: $(GO_FILES)
-	docker run --rm --user "$(shell id -u):$(shell id -g)" -v "$(shell pwd)":/usr/src/myapp -w /usr/src/myapp -e GOCACHE=/usr/src/myapp/.cache golang:1.20 go build -o perf .
+	docker run --rm --user "$(shell id -u):$(shell id -g)" -v "$(shell pwd)":/usr/src/myapp -w /usr/src/myapp -e GOCACHE=/usr/src/myapp/.cache golang:$(shell awk '/^go [0-9]+(\.[0-9]+)?$$/ {print $$2}' go.mod) go build -o perf .
```
Comment (galargh) [inline perf/impl/go-libp2p/v0.28/Makefile:6]: By getting the version from `go.mod` we don't have to update the Makefile each time.

File: perf/impl/go-libp2p/v0.28/Dockerfile
Comment (galargh) [inline perf/impl/go-libp2p/v0.28/Dockerfile:1]: I couldn't find where this Dockerfile is used either.

File: .github/workflows/add-new-impl-versions.yml (hunk 1 above)
Comment (marten-seemann) [reply .github/workflows/add-new-impl-versions.yml:4]: Schedule sounds reasonable. Should be sufficient to run this once a day.

File: .github/workflows/add-new-impl-versions.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          go-version: stable
+      - name: Get the latest version (local)
+        id: local
+        run: ls -d v* | sort -V | tail -n-1 | xargs -I{} echo "version={}" | tee -a $GITHUB_OUTPUT
+        working-directory: ${{ env.DIR }}
+      - name: Get the latest version (remote)
+        id: remote
+        env:
+          GITHUB_TOKEN: ${{ github.token }}
+        run: gh api repos/$REPO/releases/latest --jq '.tag_name' | xargs -I{} echo "version={}" | tee -a $GITHUB_OUTPUT
```
Comment (marten-seemann) [inline .github/workflows/add-new-impl-versions.yml:42]: Have you considered using `go list -m -versions github.com/libp2p/go-libp2p@latest`?

**Review by marten-seemann — COMMENTED** [review]
Very nice work @galargh! Thank you!!

File: perf/impl/go-libp2p/v0.28/.gitignore (hunk 2 above)
Comment (mxinden) [reply perf/impl/go-libp2p/v0.28/.gitignore:3]: This is created when one runs `go build .`. That said, fine to remove it.

File: perf/impl/go-libp2p/v0.28/Dockerfile
Comment (mxinden) [reply perf/impl/go-libp2p/v0.28/Dockerfile:1]: A left-over from when we were not only building but also running inside Docker. Thank you for removing.

File: .github/workflows/add-new-impl-versions.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+          cd $majorMinorRemoteVersion
+          sed -i "1s/$LOCAL_VERSION/$majorMinorRemoteVersion/g" go.mod
+          go mod tidy -go=$majorMinorGoVersion
+          go mod tidy
+          go get github.com/libp2p/go-libp2p@$REMOTE_VERSION
+          git add .
+          git commit -m "chore: add go-libp2p@$REMOTE_VERSION to $DIR"
+          git push origin $BRANCH --force
+          gh pr create --title "chore: add go-libp2p@$REMOTE_VERSION to $DIR" --body "This PR adds go-libp2p@$REMOTE_VERSION to $DIR" --head $BRANCH --base $GITHUB_REF
+          gh workflow run perf.yml --ref $BRANCH
```
Comment (mxinden) [inline .github/workflows/add-new-impl-versions.yml:63]: :pray:

**Review by mxinden — COMMENTED** [review]
Thank you Piotr! Ready to merge from my end.

File: .github/workflows/add-new-impl-versions.yml (hunk 1 above)
Comment (galargh) [reply .github/workflows/add-new-impl-versions.yml:4]:
```suggestion
  workflow_dispatch:
  schedule:
    # “At minute 36 past every 24th hour.” - https://crontab.guru/#36_*/24_*_*_*
    - cron:  '36 */24 * * *'
```

**Comment by BigLep** [comment]
Sweet!  What's needed to merge this?
@mxinden : will you create something similar for rust-libp2p and js-libp2p?

**Review by mxinden — APPROVED** [review]
Ready to merge from my end.

**Comment by mxinden** [comment]
> @mxinden : will you create something similar for rust-libp2p and js-libp2p?

Ideally. That said, it is not high on my priority list. E.g. I am prioritizing https://github.com/libp2p/test-plans/issues/261

**Pull request merged**